_posts/2017-02-09-whats-bonnyci-wheres-it-going.markdown

```diff
@@ -0,0 +1,170 @@
+nobody is all that bothered that this takes 4 * *testduration* to land
+all 4 commits. Jenkins will spin up one node at a time, run tests,
+and merge code, and then do the same again, stacking each one on top
+of master. If one fails, it will naturally be skipped, and life will go
+on. If the tests take an hour to run, this takes 4 hours.
+
+However, as the team grows, a problem arises. Lets say the team grows to
+10 contributors, and sticks with the one reviewer policy. Now the potential
+for a clogged gate is much higher. It's likely a reviewer will have many more
+patches available to them to approve. If a reviewer now approves a patch from
+each of their fellow contributors, so 9 patches, they will be facing _9 hours_
+before they know if any of these patches interact negatively with any of the
+others. This compounds as failures happen. If the second patch to land introduces
+an API breakage in which its tests pass, but many other new changes that rely
+on the old behavior break, suddenly you may only land 1 patch in 9 hours.
+
+While the JClouds plugin helps with pre-approval checking parallelism,
+it can do nothing with this problem.
+
+This is why Zuul implements speculative futures. In the 9 patch scenario above,
+Zuul will spin up _9_ cloud nodes all at once. It will try to run a job for
+each potential future in the master branch. So if the patches are numbered, it will
+run each of these futures in parallel:
+
+
+    A 1
+    B 1+2
+    C 1+2+3
+    D 1+2+3+4
+    E 1+2+3+4+5
+    F 1+2+3+4+5+6
+    G 1+2+3+4+5+6+7
+    H 1+2+3+4+5+6+7+8
+    I 1+2+3+4+5+6+7+8+9
+
+If patch #2 introduces an API break that breaks #3 and #9, then This will happen:
+
+    (A) merges #1
+    (B) merges #2
+    (C) fails
+    (D - I) are "reset" and their results are ignored by Zuul
+
+Now (C) will be reported as a failure, and Zuul will try a new scenario:
+
+    AA 4
+    BB 4+5
+    CC 4+5+6
+    DD 4+5+6+7
+    EE 4+5+6+7+8
+    FF 4+5+6+7+8+9
+
+And this happens:
+
+    (AA) merges #4
+    (BB) merges #5
+    (CC) merges #6
+    (DD) merges #7
+    (EE) merges #8
+    (FF) fails
+
+So, in the time span of _two_ test runs, we have found the two
+bad interactions, and merged 7 patches that do work well all
+together. Suddenly having hour long tests isn't such a huge problem.
+
+Meanwhile in those two test runs, new patches have only been created from
+known good working master branches, thus reinforcing good interactions
+with approved code, allowing development to progress faster and reducing
+accidental breakage. Even at the busiest time, usually just before
+releases, OpenStack sees a very long gate, and exhaustion of its 100's
+of available VM capacity across many clouds. But it still doesn't usually
+get worse than about 8 or 9 test runs to land many patches.
+
+This becomes especially important when you want to start co-gating
+projects.  In OpenStack, we don't want to merge code to Keystone that
```

Comment:
Good feedback. I will try using the service names rather than project names. People will understand "merge code to the identity and authentication service that breaks the compute service".